source/azure_iot.c

```diff
@@ -0,0 +1,455 @@
+        ( ppucOutValue == NULL ) || ( pulOutValueLength == NULL ) )
+    {
+        AZLogError( ( "AzureIoT_MessagePropertiesFind failed: Invalid argument" ) );
+        return eAzureIoTInvalidArgument;
+    }
+
+    xResult = az_iot_message_properties_find( &pxMessageProperties->_internal.xProperties,
+                                              xNameSpan, &xOutValueSpan );
+
+    if( az_result_failed( xResult ) )
+    {
+        return eAzureIoTItemNotFound;
+    }
+
+    *ppucOutValue = az_span_ptr( xOutValueSpan );
+    *pulOutValueLength = ( uint32_t ) az_span_size( xOutValueSpan );
+
+    return eAzureIoTSuccess;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTResult_t AzureIoT_Base64HMACCalculate( AzureIoTGetHMACFunc_t xAzureIoTHMACFunction,
+                                               const uint8_t * pucKey,
+                                               uint32_t ulKeySize,
+                                               const uint8_t * pucMessage,
+                                               uint32_t ulMessageSize,
+                                               uint8_t * pucBuffer,
+                                               uint32_t ulBufferLength,
+                                               uint8_t * pucOutput,
+                                               uint32_t ulOutputSize,
+                                               uint32_t * pulOutputLength )
+{
+    AzureIoTResult_t xStatus;
+    uint8_t * pucHashBuf;
+    uint8_t * pucDecodedKeyBuf = pucBuffer;
+    uint32_t ulHashBufSize = azureiotBASE64_HASH_BUFFER_SIZE;
+    uint32_t ulBinaryKeyBufSize;
+
+    if( ( xAzureIoTHMACFunction == NULL ) ||
+        ( pucKey == NULL ) || ( ulKeySize == 0 ) ||
+        ( pucMessage == NULL ) || ( ulMessageSize == 0 ) ||
+        ( pucBuffer == NULL ) || ( ulBufferLength == 0 ) ||
+        ( pucOutput == NULL ) || ( pulOutputLength == NULL ) )
+    {
+        AZLogError( ( "AzureIoT_Base64HMACCalculate failed: Invalid argument" ) );
+        return eAzureIoTInvalidArgument;
+    }
+
+    xStatus = prvAzureIoTBase64Decode( ( char * ) pucKey, ulKeySize,
+                                       pucDecodedKeyBuf, ulBufferLength, &ulBinaryKeyBufSize );
+
+    if( xStatus )
+    {
+        return xStatus;
+    }
+
+    /* Decoded key is less than total decoded buffer size */
+    ulBufferLength -= ulBinaryKeyBufSize;
+
+    if( ulHashBufSize > ulBufferLength )
+    {
+        return eAzureIoTOutOfMemory;
+    }
+
+    pucHashBuf = pucDecodedKeyBuf + ulBinaryKeyBufSize;
+    memset( pucHashBuf, 0, ulHashBufSize );
+
+    if( xAzureIoTHMACFunction( pucDecodedKeyBuf, ulBinaryKeyBufSize,
+                               pucMessage, ( uint32_t ) ulMessageSize,
+                               pucHashBuf, ulHashBufSize, &ulHashBufSize ) )
+    {
+        return eAzureIoTFailed;
+    }
+
+    uint32_t ulBase64OutputLength;
+    xStatus = prvAzureIoTBase64Encode( pucHashBuf, ulHashBufSize,
```

Comment:
lets define all variable at start of function. It really help in code review and debugging to know how much stack, function is using. (more closer to what compiler will do). Similar to AZRTOS

Reply:
and FreeRTOS too.

---
In reply to: [608244686](https://github.com/Azure/azure-iot-middleware-freertos/pull/58#discussion_r608244686) [](ancestors = 608244686)

Reply:
yea usually I would say this is a religious debate but if it's similar to the others then I'm good with it